src/gameboard.js

```diff
@@ -0,0 +1,149 @@
+import shipFactory from './ship';
+import { randomBoolean, randomCoordinates } from './utilities';
+
+const gameboardFactory = () => {
+  let matrix = [];
+  matrix.length = 10;
+  for (let n = 0; n < 10; n++) {
+    matrix[n] = new Array(10).fill(' ');
+  }
+
+  const validCoordinates = (l, isHorizontal, coordinates) => {
+    if (
+      matrix[coordinates[0]][coordinates[1]] !== ' ' ||
+      (isHorizontal && coordinates[1] + l > 10) ||
+      (!isHorizontal && coordinates[0] + l > 10)
+    )
+      return false;
+
+    const row = coordinates[0];
+    const col = coordinates[1];
+
+    let rowStart = row - 1;
+    if (row === 0) rowStart = 0;
+    let colStart = col - 1;
+    if (col === 0) colStart = 0;
+
+    let rowEnd, colEnd;
+
+    if (isHorizontal) {
+      rowEnd = row + 2;
+      if (row === 9) rowEnd = 10;
+      colEnd = col + l + 1;
+      if (col + l === 10) colEnd = 10;
+    } else {
+      rowEnd = row + l + 1;
+      if (row + l === 10) rowEnd = 10;
+      colEnd = col + 2;
+      if (col === 9) colEnd = 10;
+    }
+
+    for (let i = rowStart; i < rowEnd; i++) {
+      for (let j = colStart; j < colEnd; j++) {
+        if (matrix[i][j] !== ' ') return false;
+      }
+    }
+    return true;
+  };
+
+  const placeShip = (l, isHorizontal, coordinates) => {
+    if (!validCoordinates(l, isHorizontal, coordinates)) return -1;
+
+    const ship = shipFactory(l, isHorizontal);
```

Comment:
This `shipFactory` usage is a tough one. The better (and actually the more proper) way to do this is pass an "instantiated" `ship` object as argument to `gameBoar`. To be able to test `gameBoard` more properly (i.e, we wanna make sure that whenever we call the `gameBoard.receiveAttack(x,y)`, for instance, that the `ship` 's `hit` method was called:

```
// We can then mock the ship object and the corresponding hit method like this
const ship = {hit: jest.fn()}

gameBoard.receiveAttack(0,0);

expect(ship.hit).toHaveBeenLastCalled();
```

We can only easily do this kind of test if we are doing "dependency" injection rather than instantiating the ship object right inside the `gameBoard`.